API Documentation Improvements

seldridge opened the pull request:
<!--
Please select the item best describing the pull request in each category and delete the other items.
-->
**Related issue**: #872 

<!-- choose one -->
**Type of change**: other enhancement

<!-- choose one -->
**Impact**: no functional change

<!-- choose one -->
**Development Phase**: proposal

This migrates all `do_*` methods into a dedicated ScalaDoc group: `SourceInfoTransformMacro`. This is done using a bare trait to have inheritance propogate the group description (as opposed to a ScalaDoc macro).

This also adds API documentation and additional groups: `Arithmetic`, `Comparison`, and `Connect`. 

~This includes two small (apparent?) bug fixes where both `SInt` and `FixedPoint` were missing their `do_unary-` and `do_unary-%` methods. These were provided as `unary-` and `unary-%`. I _think_ this was a mistake, but I'm not entirely sure why it wasn't tripping anything.~

### Todo
- [x] Finish documentation everything in `Bits.scala`

**Release Notes**
<!--
Text from here to the end of the body will be considered for inclusion in the release notes for the version containing this pull request.
-->

- Improve API documentation using ScalaDoc groups

seldridge commented:
While going through this, some other refactoring items became apparent:
  - `Bits.scala` is just a massive. This should likely be split into `{SInt, UInt, Bool, Bits, FixedPoint}` files.
  - All the `do_*` could be moved into self-typed traits. This is purely cosmetic, but it would improve code readability, IMO. I had thought that this could be combined with scaladoc's `-skip-packages` option to completely remove the `do_*` methods from the output scaladoc. However, that skips classes/objects in the package and not methods inherited.
  - The discussion of moving more methods into `Num` could be revisited. There's a lot of duplicate documentation that stems from `Num` not providing more abstract methods.

ducky64 commented:
I believe the reason Bits is massive is because the sealed class hierarchy requires everything to be in one file. I don't think this is the only file affected by that limitation either...

seldridge commented:
Somehow I missed that. Makes sense, though. Refactoring into traits would circumvent that somewhat. There may be some other patterns that make sense here. Low priority, though.

seldridge commented:
Note: I rebased this to provide a logical order. However, [GitHub is apparently smarter than I am and knows that a chronological order is best because reasons](https://help.github.com/articles/why-are-my-commits-in-the-wrong-order/). Looking at the original branch shows things correctly: https://github.com/freechipsproject/chisel3/compare/master...seldridge:issue-872. However, the order of commits above is mixed up with the Travis :heavy_check_mark: showing up in the middle of the patch sequence even though that's the last commit.

Something to be aware of as it relates to the discussion about merge vs. squash and asking users to rebase commits into a sane order before merging. There would be no problem here if I had rebased before PR'ing.

jackkoenig commented on chiselFrontend/src/main/scala/chisel3/core/Bits.scala:
```diff
@@ -214,70 +240,135 @@ sealed abstract class Bits(width: Width)
+    * @return this component with `that` many zeros concatenated to its
+    * least significant end.
+    *
+    * @note the width of the output is `that + this.getWidth`.
```
I think `width of this` is also better

seldridge commented on chiselFrontend/src/main/scala/chisel3/core/Bits.scala:
```diff
@@ -99,9 +102,27 @@ sealed abstract class Bits(width: Width)
+    *
+    * @param n the number of bits to remove
+    *
+    * @return This component with the `n` most significant bits removed.
```
The spaces were for clarity and to align with what I thought scala mainline was doing, but I didn't remember correctly. I'm not aware of any formal guidelines for this and scala mainline isn't even consistent within the same file (e.g., in https://github.com/scala/scala/blob/2.13.x/src/library/scala/collection/immutable/List.scala#L81 things sometimes have a space, sometimes they're aligned, etc.). I'll try to follow https://docs.scala-lang.org/style/scaladoc.html. Summarily:
  - One empty line between the title/descriptoin and the first `@` tags
  - Indent things to align under their parent (when needed)
  - Use some of the conventions for talking about classes and objects
  - Figure out what can be a macro here

seldridge commented on coreMacros/src/main/scala/chisel3/internal/sourceinfo/SourceInfoDoc.scala:
```diff
@@ -0,0 +1,22 @@
+  *
+  * @groupprio SourceInfoTransformMacro 1001
+  */
+trait SourceInfoDoc
```
I'd have to think about it. Maybe broader discussion. I think I'm of the opinion that if something is a public API, then it should be in the ScalaDoc. I can see hiding it if and only if no other public thing would generate or want to generate a reference into it. (There's a secondary reason that @ducky64 has some _really_ good documentation in internal and I'm currently referencing that...)

Granted, the internal stuff in question is a lot of macro magic. I'm not sure if that clutters things up. Maybe worth a discussion as a blanket policy.

seldridge commented on chiselFrontend/src/main/scala/chisel3/core/Bits.scala:
```diff
@@ -464,10 +671,27 @@ sealed class UInt private[core] (width: Width) extends Bits(width) with Num[UInt
     new UInt(w).asInstanceOf[this.type]
 
   // TODO: refactor to share documentation with Num or add independent scaladoc
+  /** Unary negation (increasing width)
```
Sounds better to me. Increasing sounds like `+1` while expanding seems more like the width is widening. This is likely a good candidate for a macro.

seldridge commented:
I updated this to address your comments @jackkoenig. I think I got all the `do_*` and other, similar methods that don't have the `do_` prefix.

I realized that there was way too much boilerplate and started abusing ScalaDoc macros. It helps immensely to be consistent. I'm using a convention from Scala's collections where a macro of `$coll` is defined as the current collection you're documenting.

I did run into this (https://issues.scala-lang.org/browse/SI-9785): ScalaDoc doesn't properly substitute for macros used in parent methods that are overriding by macros in children. It's not the end of the world, but it would make some of the `Num` methods look a little prettier (they currently say `a hardware type` when they could say, e.g., `a [[UInt]]`).

seldridge commented:
Updated to ignore chisel3.internal. This required moving `SourceInfoDoc` out of internal. This is now logically in `chisel3`, but physically in `coreMacros/src/main/scala/chisel3`. Unfortunately, if I make that `private [chisel3]` the group description is ignore and the `Chisel` compatibility package can't see it (as that's external to `chisel3`.

jackkoenig commented on chiselFrontend/src/main/scala/chisel3/core/Bits.scala:
```diff
@@ -214,96 +258,134 @@ sealed abstract class Bits(width: Width)
   final def toBools(): Seq[Bool] = macro SourceInfoTransform.noArg
 
+  /** @group SourceInfoTransformMacro */
   def toBools(implicit sourceInfo: SourceInfo, compileOptions: CompileOptions): Seq[Bool] =
```
It's really weird that this one doesn't have a `do_` prefix, @ducky64 any idea? I'm kind of surprised this even works...

jackkoenig commented on coreMacros/src/main/scala/chisel3/SourceInfoDoc.scala:
```diff
@@ -0,0 +1,42 @@
+  *   <br>
+  *   <br>
+  *
+  *   🐉🐉🐉 '''Here be dragons...''' 🐉🐉🐉
```
Nice dragons :)
